### PR TITLE
perf: harden homogeneous decode batching

### LINKED ISSUE #1642: [Gemma E4B Perf] Root: close release Melix serving gap vs OMLX/SwiftLM
> *This was generated by AI during triage.*

    ## Triage

    - Category: bug
    - State: ready-for-human (encoded in body; this repository currently has no matching triage-state labels)
    - Type: Tracker
    - Scope level: Root

    ## Parent

    None - root issue.

    ## What to build

    Melix release binaries still lag OMLX/SwiftLM on `models--unsloth--gemma-4-E4B-it-MLX-8bit` after the debug-build measurement error was corrected.

Benchmark scope: warm profile, usage-token accounting, prompt targets 128/1024, max output 64, concurrency 1/2, repeats 3, 54 observations, 0 errors.

| Prompt | Conc | Best total | Melix total | Melix total gap | Melix decode | Best decode | Decode gap |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 128 | 1 | SwiftLM 1739.0 ms | 1878.8 ms | +8.0% | 39.1 tok/s | SwiftLM 43.0 tok/s | -9.2% |
| 128 | 2 | OMLX 1458.5 ms | 2579.0 ms | +76.8% | 29.7 tok/s | OMLX 56.0 tok/s | -47.1% |
| 1024 | 1 | OMLX 1438.7 ms | 2202.1 ms | +53.1% | 35.9 tok/s | SwiftLM 66.7 tok/s | -46.1% |
| 1024 | 2 | OMLX 1607.9 ms | 2437.5 ms | +51.6% | 35.9 tok/s | OMLX 71.1 tok/s | -49.6% |

Confirmed causes and non-causes from the current evidence:
- The first comparison was contaminated by debug Melix binaries; the corrected run used `.build/release` Swift binaries and improved Melix total latency by 17.1-44.8% versus debug.
- Remaining gap is concentrated in decode throughput and concurrency 2 behavior, not memory pressure: host memory had 94% free and Melix text-worker RSS was 7.807 GiB, comparable to SwiftLM and lower than OMLX.
- Melix reports `scheduler.continuous_batch_size=2` and `continuous_batch_merge_rate=31.6%`, but worker code still processes a single `GenerateRequest` stream at a time; the current evidence does not prove model-level fused decode batching.
- Acceleration/profile counters were effectively baseline for this run: active KV, sparse prefill, accelerated prefill, speculative decode, and dFlash counters were all zero or disabled.

Evidence bundle: `.runtime/serving-comparison/gemma-e4b-20260528/threeway-gemma-e4b-release-warm-usage-20260528T171827JST`.

    ## Acceptance criteria

    - [ ] A release Melix rerun on the same model/profile has no scenario more than 25% slower than the best peer median total latency.
- [ ] A release Melix rerun has no scenario more than 25% below the best peer median decode tok/s.
- [ ] The benchmark artifact records Melix binary paths, binary hashes, build mode, revisions, token sources, measurement profile, and per-request phase attribution.
- [ ] The issue tree below is resolved or superseded by narrower tickets with equivalent acceptance criteria.

    ## Blocked by

    - None - can start immediately.

## Related Issues

- #1460
- #1461
- #1392
- #1396
- #350

## Evidence and Notes

- Related existing performance work includes #1460, #1461, #1392, #1396, and #350; this issue tree is scoped to the current Gemma E4B text-serving release benchmark gap.
- Do not treat the earlier debug-Melix run as a fair runtime comparison; use it only as build-mode evidence.

### LINKED ISSUE #1654: [Gemma E4B Perf P2.2] Implement homogeneous Swift text batch decode path
> *This was generated by AI during triage.*

    ## Triage

    - Category: enhancement
    - State: ready-for-human (encoded in body; this repository currently has no matching triage-state labels)
    - Type: Tracker
    - Scope level: Plan

    ## Parent

    #1650

    ## What to build

    Introduce a narrow model-level batch decode path for homogeneous Swift text requests that share model, acceleration profile, sampling-compatible settings, and compatible prompt/decode constraints, with fallback to the existing single-request path for everything else.

    ## Acceptance criteria

    - [ ] U2.2.1 and U2.2.2 are closed.
- [ ] Concurrency 2 Gemma E4B release benchmark improves aggregate output tok/s and total latency versus the current release baseline.
- [ ] Batching does not break per-request streaming order, usage accounting, cancellation, or error isolation.

    ## Blocked by

    - #1651

### LINKED ISSUE #1656: [Gemma E4B Perf U2.2.2] Preserve cancellation, fallback, and stream ordering in batch decode
> *This was generated by AI during triage.*

    ## Triage

    - Category: enhancement
    - State: ready-for-agent (encoded in body; this repository currently has no matching triage-state labels)
    - Type: AFK
    - Scope level: Unit

    ## Parent

    #1654

    ## What to build

    Harden the batch decode path so incompatible requests, cancellation, early finish, model errors, and client disconnects fall back or unwind without corrupting other requests in the batch.

    ## Acceptance criteria

    - [ ] Tests cover cancellation of one request while another continues.
- [ ] Tests cover incompatible sampling/model requests falling back to the single-request path.
- [ ] SSE/OpenAI event order and usage accounting match the existing single-request contract.

    ## Blocked by

    - #1655

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Closes #1656 by hardening homogeneous Swift text batch decode fallback, cancellation, stream ordering, and usage-accounting coverage.
- Supports #1654 and #1642 by adding scheduler cohort annotations, worker-side cohort decode batching, cache-compatibility checks, batch decode metrics, and preserved three-way benchmark run evidence.
- This does not close #1654 or #1642: the current release three-way recompute still misses the #1642 25% gates for three scenarios.

## Plan or Spec
- `docs/plans/2026-05-29-issue-1655-homogeneous-decode-coordinator.md`
- `docs/adr/0001-homogeneous-batch-decode-cache-compatibility.md`

## Commands Run
```text
make bootstrap
# Passed. Installed repository git hooks and completed locked Python sync checks.

make proto
# Passed. No generated protobuf diffs.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python --extra mlx coverage run -m pytest -q tests/test_three_way_serving_compare.py tests/test_omlx_melix_compare_benchmark.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python --extra mlx coverage report --include 'tests/test_three_way_serving_compare.py,tests/test_omlx_melix_compare_benchmark.py'
uv run python scripts/changed_scope_coverage.py --coverage-json .runtime/coverage/benchmark-tests.json --diff-base origin/main --path tests/test_three_way_serving_compare.py --path tests/test_omlx_melix_compare_benchmark.py
# Passed. 100 tests passed. Changed-line coverage: 66/66 = 100%.

swift test --package-path services/control-plane-swift --enable-code-coverage --filter 'EventSubscriptionHubTests|RequestCoordinatorTests' --disable-automatic-resolution
scripts/swift_changed_line_coverage.py --package-path services/control-plane-swift --diff-base origin/main
# Passed. 100 tests passed. Changed-line coverage: 250/257 = 97.28%.

swift test --package-path services/mlx-text-worker-swift --enable-code-coverage --filter WorkerScaffoldTests
scripts/swift_changed_line_coverage.py --package-path services/mlx-text-worker-swift --diff-base origin/main
# Passed. 219 tests passed. Changed-line coverage: 2165/2215 = 97.74%.

git diff --check
# Passed.

git commit -m "perf: harden homogeneous decode batching"
# Passed pre-commit full gate on a 256 GiB macOS host:
# - make swift-test: passed. Protocol 1 test, text worker 219 tests, control-plane suites including worker clients 114 tests, macOS menu bar 779 tests.
# - make py-test: 3373 passed, 14 skipped, 2 warnings in 148.50s.
# - make integration-test: 115 passed, 1 skipped in 662.21s.
# - PR-scoped performance probe same-cohort-batching-probe-evidence: status ok, regressions 0, verification failures 0.

Three-way benchmark evidence:
# Artifact:
# .runtime/serving-comparison/gemma-e4b-20260601-latest-ae720be-rerun9/threeway/gemma-e4b-mainae720be-release-threeway-20260601-rerun9-full
# Saved run-evidence.json, acceptance-25pct-recompute.json, and acceptance-25pct-recompute.md.
# Recomputed #1642 25% acceptance: threshold_failed, 6 failures.
# - 128/c1: total +32.22% fail, decode -27.43% fail
# - 128/c2: total +41.70% fail, decode -34.33% fail
# - 1024/c1: total +44.96% fail, decode -37.17% fail
# - 1024/c2: total +7.00% ok, decode -1.32% ok
```

## Coverage and Metrics
- Python changed-line coverage: 100% (66/66).
- Control-plane changed-line coverage: 97.28% (250/257).
- Swift text worker changed-line coverage: 97.74% (2165/2215).
- Pre-commit scoped performance report: `.runtime/pre-commit-performance/20260531-200729-5c1f1115/report/report.md`.
- Pre-commit same-cohort batching probe: status ok, no regressions, no verification failures. Both base and head report `status_warning=1` because the deterministic probe still observes scheduler cohort size 2 and worker model/decode batch size 1; this is not a regression, and the new worker tests cover the real batch, fallback, cancellation, ordering, and usage-accounting paths.
- Observability mode: minimal runtime metrics plus evidence-mode benchmark artifacts. The full-logits model-eval synchronization probe is opt-in via `MELIX_SWIFT_BATCH_DECODE_FORCE_MODEL_EVAL_PROBE=1`; the default hot path does not force that sync.
- Probe overhead: the pre-commit base/head same-cohort probe took about 118-121 seconds per side in isolated temp worktrees.
- #1642 release benchmark metric: current recompute failed the 25% total-latency/decode-throughput gates for 128/c1, 128/c2, and 1024/c1. The successful 1024/c2 row is not sufficient to satisfy #1642.

## Known Gaps
- Does not close #1642. The 25% release acceptance still fails for 128/c1, 128/c2, and 1024/c1.
- Does not close #1654. The parent homogeneous decode performance issue still needs benchmark improvement evidence and this #1656 slice merged/closed.
- The same-cohort PR-scoped probe still reports `status_warning=1` for the deterministic scheduler/worker batch-size mismatch; it is stable across base/head and covered by focused worker tests in this PR.
- The three-way benchmark artifact reflects current host state and is diagnostic evidence, not #1642 closure evidence.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts. N/A: no protocol changes; `make proto` produced no diff.
- [x] Dependency changes include updated lockfiles. N/A: no dependency changes.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-text-worker-swift/Sources/Core/Runtime/SwiftMLXBackend.swift [hunk 1]
```diff
@@ -1139,6 +1150,731 @@ private func makePreparedSpeculativeDecodeEvents(
...
+    decodeTokenIDCallCount += 1
+    decodeTokenIDTotalMicros += elapsedMicros(since: tokenIDStartedAt)
+
+    for (batchIndex, requestIndex) in activeIndices.enumerated() {
+        let tokenID = ids[batchIndex]
+        states[requestIndex].output = LMOutput(
+            logits: logits[batchIndex ..< (batchIndex + 1), 0..., 0...]
+        )
+        states[requestIndex].pendingToken = MLXArray([tokenID])
+        states[requestIndex].pendingTokenID = tokenID
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-text-worker-swift/Sources/Core/Runtime/SwiftMLXBackend.swift:1707]: Addressed in follow-up PR #1714. I verified current MLX returns `.uint32` for this `argMax` path, not `.int32`, and that `asArray(Int.self)` converts through MLX in the current dependency rather than raw-stride casting. The follow-up still hardens the hot path by reading `UInt32` explicitly and converting to `Int`, with `testAutoSwiftMLXBackendBatchArgMaxTokenIDsBridgeThroughUInt32` covering the bridge.

File: services/mlx-text-worker-swift/Sources/Core/Runtime/SwiftMLXBackend.swift [hunk 2]
```diff
@@ -1139,6 +1150,731 @@ private func makePreparedSpeculativeDecodeEvents(
...
+        split = RotatingKVCache(maxSize: max(1, maxSize))
+    } else {
+        let simple = KVCacheSimple()
+        if let sourceStep = adapter?.sourceSimpleStep {
+            simple.step = sourceStep
+        } else if let source = layer as? KVCacheSimple {
+            simple.step = source.step
+        }
+        split = simple
+    }
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-text-worker-swift/Sources/Core/Runtime/SwiftMLXBackend.swift:1869]: Addressed in follow-up PR #1714. The adapter now exposes its underlying `KVCache` to this file, and `splitBatchDecodeCacheLayer` checks the underlying layer for `RotatingKVCache` instead of relying on `metaState.count`. Added `testAutoSwiftMLXBackendBatchDecodeSplitsRotatingAdapterByUnderlyingCacheType` for the adapter-wrapped rotating cache path.
